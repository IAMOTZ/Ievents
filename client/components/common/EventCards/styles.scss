@import "../../../styles/variables";

.event-card {
  border-color: $custom-blue;
  .card-header {
    border-bottom-color: $custom-blue;
  }
  .card-body {
    .description {
      height: 75px;
    }
    .read-more-link {
      font-size: 13px;
    }
    .date-and-status {
      font-size: 15px;
      display: flex;
      justify-content: space-between;
    }
  }
  .card-footer {
    border-top-color: $custom-blue;
    .action-links {
      display: flex;
      justify-content: space-between;
      a, span {
        color: #343a40;
        font-size: 15px;
      }
    }
  }
}
