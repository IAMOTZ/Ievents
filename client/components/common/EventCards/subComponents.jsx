import React from 'react';
import PropTypes from 'prop-types';
import { Link } from 'react-router-dom';

/**
 * An helper function to truncate a text.
 * @param {String} text The text to truncate.
 * @returns {String} The truncated text.
 */
const truncateText = (text) => {
  let newText;
  if (text && text.length > 100) {
    newText = text.substr(0, 70);
    newText += '...';
  }
  return newText;
};

/* eslint-disable react/forbid-prop-types */
export const RenderEventDescription = (props) => {
  const { event } = props;
  const truncatedText = truncateText(event.description);
  return (
    <div className="description">
      {
        truncatedText ?
          <span>
            <span>{truncatedText}</span>
            <span className="read-more-link">
              &nbsp;
              <a
                className="no-outline-btn"
                data-toggle="modal"
                href="#event-details-modal"
                onClick={() => props.createModalContent(event)}
                role="button"
              >Read more
              </a>
            </span>
          </span> :
          <span>{event.description}</span>
      }
    </div>
  );
};
RenderEventDescription.propTypes = {
  createModalContent: PropTypes.func.isRequired,
  event: PropTypes.object.isRequired,
};

export const RenderEventFooter = (props) => {
  const { event } = props;
  return (
    <div className="card-footer">
      <div className="action-links">
        <span>
          {
            event.status === 'allowed' ?
              <Link to={`/events/${event.id}/edit`} onClick={props.edit} id={event.id} >
                <i className="fa fa-pencil fa-fw text-dark" />
                <span>Edit Event</span>
              </Link> :
              <span>
                {
                  event.status === 'canceled' ?
                    <i className="fa fa-times text-danger" /> :
                    <i className="fa fa-check text-success" />
                }
                &nbsp;<span className="text-capitalize">{event.status}</span>
              </span>
          }
        </span>
        <span>
          <a href="#confirmation-modal" data-toggle="modal" onClick={props.startDelete} id={event.id}>
            <i className="fa fa-trash text-dark" />&nbsp;
            <span>Delete Event</span>
          </a>
        </span>
      </div>
    </div>
  );
};
RenderEventFooter.propTypes = {
  edit: PropTypes.func.isRequired,
  event: PropTypes.object.isRequired,
  startDelete: PropTypes.func.isRequired,
};

export const RenderEventBody = (props) => {
  const { event } = props;
  return (
    <div className="card-body">
      <div className={event.status === 'allowed' ? '' : 'text-muted'}>
        <RenderEventDescription
          event={event}
          createModalContent={props.createModalContent}
        />
        <p className="mb-1">
          <i className="fa fa-bank fa-fw" />&nbsp;
          <span className="text-capitalize">{event.centerName}</span>
        </p>
        <div className="date-and-status">
          <span>{event.date.replace(/-/g, '/')}&nbsp;</span>
        </div>
      </div>
    </div>
  );
};
RenderEventBody.propTypes = {
  event: PropTypes.object.isRequired,
  createModalContent: PropTypes.func.isRequired,
};

export const RenderEventBodyInTransactions = (props) => {
  const { event } = props;
  return (
    <div className="card-body">
      <RenderEventDescription
        event={event}
        createModalContent={props.createModalContent}
      />
      <p className="mb-1">
        <i className="fa fa-user fa-fw" />&nbsp;
        <span>{event.user.email}</span>
      </p>
      <div className="date-and-status">
        <span>{event.date.replace(/-/g, '/')}&nbsp;</span>
      </div>
      <button
        onClick={props.startEventCancel}
        data-toggle="modal"
        data-target="#confirmation-modal"
        id={event.id}
        className="btn ie-red-button mt-2 btn-block"
      >
        {
          props.cancelingTransactionStarted ?
            <span>Canceling&nbsp;<i className="fa fa-spinner fa-pulse" /></span> : <span>Cancel Event</span>
        }
      </button>
    </div>
  );
};
RenderEventBodyInTransactions.propTypes = {
  createModalContent: PropTypes.func.isRequired,
  event: PropTypes.object.isRequired,
  startEventCancel: PropTypes.func.isRequired,
  cancelingTransactionStarted: PropTypes.bool.isRequired,
};
