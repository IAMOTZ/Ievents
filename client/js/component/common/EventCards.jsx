import React from 'react';

export default (props) => {
  const eventCards = props.events.map((event => {
    return (
      <div className="card text-dark bg-white mb-3 mx-auto" style={{ maxWidth: 20 + 'rem' }} key={event.id}>
        <div className="card-header h4">
          {event.title}
          <a href="" className="ml-4 mr-2">
            <i className="fa fa-pencil"></i>
          </a>
          <a href="#" className="text-primary">
            <i className="fa fa-trash" onClick={props.startDelete} id={event.id}></i>
          </a>
        </div>
        <div className="card-body">
          <p className="card-text">{event.description}</p>
          <span className="text-muted">{event.date}&nbsp;</span>
          <a href="#" className="btn btn-block bg-dark text-white mt-2" onClick={props.remove} id={event.id}>Done</a>
        </div>
      </div>
    );
  }));
  return eventCards;

}