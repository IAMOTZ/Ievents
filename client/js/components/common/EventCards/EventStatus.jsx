import React from 'react';
import PropTypes from 'prop-types';

const EventStatus = (props) => {
  switch (props.status) {
    case ('canceled'): {
      return (
        <div className="card-footer text-danger text-center">
          <i className="fa fa-times-circle" />&nbsp;
          <span>Canceled</span>
        </div>
      );
    }
    case ('done'): {
      return (
        <div className="card-footer text-info text-center">
          <i className="fa fa-exclamation" />&nbsp;
          <span>Done</span>
        </div>
      );
    }
    default: {
      return (
        <div className="card-footer text-success text-center">
          <i className="fa fa-check" />&nbsp;
          <span>Allowed</span>
        </div>
      );
    }
  }
};

EventStatus.defaultProps = {
  status: null,
};
EventStatus.propTypes = {
  status: PropTypes.string,
};
export default EventStatus;
