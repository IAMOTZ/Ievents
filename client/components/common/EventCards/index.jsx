import React from 'react';
import PropTypes from 'prop-types';
import './styles.scss';
import {
  RenderEventBody,
  RenderEventBodyInTransactions,
  RenderEventFooter
} from './subComponents';

/* eslint-disable react/forbid-prop-types */
const EventCards = props => props.events.map(event => (
  <div className="col-lg-4 col-md-6 col-sm-1" key={event.id}>
    <div
      style={{ maxWidth: `${20}rem` }}
      className="card text-dark bg-white mb-3 mx-auto event-card"
    >
      <div className="card-header">
        <span className="h4 mb-0">{event.title}</span>
      </div>
      {
        props.isTransactionsPage ?
          <RenderEventBodyInTransactions
            event={event}
            createModalContent={props.createModalContent}
            startEventCancel={props.startEventCancel}
            cancelingTransactionStarted={props.cancelingTransactionStarted}
          /> :
          <RenderEventBody
            event={event}
            createModalContent={props.createModalContent}
          />
      }
      {
        props.isTransactionsPage ?
          null :
          <RenderEventFooter
            event={event}
            startDelete={props.startDelete}
            edit={props.edit}
          />
      }
    </div>
  </div>
));
EventCards.propTypes = {
  events: PropTypes.array,
};
EventCards.defaultProps = {
  events: []
};


export default EventCards;
