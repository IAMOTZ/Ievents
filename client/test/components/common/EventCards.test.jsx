import React from 'react';
import EventCards from '../../../components/common/EventCards';
import {
  RenderEventBody, RenderEventBodyInTransactions,
  RenderEventDescription, RenderEventFooter
} from '../../../components/common/EventCards/subComponents';

const sampleEvent = {
  id: 1,
  centerId: 1,
  userId: 1,
  title: 'test-event',
  description: 'test description',
  date: '12/34/2020',
  status: 'allowed',
  user: {
    email: 'tester@gmail.com'
  }
};

describe('<EventCards />', () => {
  const props = {
    isTransactionsPage: false,
    events: [sampleEvent],
    edit: jest.fn(),
    cancelingTransactionStarted: false,
    createModalContent: jest.fn(),
    startEventCancel: jest.fn(),
    startDelete: jest.fn(),
  };
  const alterProps = newProps => ({ ...props, ...newProps });
  let wrapper;
  beforeEach(() => {
    wrapper = shallow(<EventCards {...props} />);
  });
  describe('Rendering:', () => {
    it('should render correctly on Events page', () => {
      expect(wrapper.first().shallow()).toMatchSnapshot();
    });
    it('should render correctly on Transactions page', () => {
      wrapper.setProps(alterProps({ isTransactionsPage: true }));
      expect(wrapper.first().shallow()).toMatchSnapshot();
    });
  });
});

describe('<RenderEventBody />', () => {
  const props = {
    event: sampleEvent,
    createModalContent: jest.fn(),
  };
  const alterProps = newProps => ({ ...props, ...newProps });
  let wrapper;
  beforeEach(() => {
    wrapper = shallow(<RenderEventBody {...props} />);
  });
  describe('Rendering:', () => {
    it('should render correctly', () => {
      expect(wrapper).toMatchSnapshot();
    });
    it('should render correctly when the event is not allowed', () => {
      wrapper.setProps(alterProps({ event: { ...sampleEvent, status: 'canceled' } }));
      expect(wrapper).toMatchSnapshot();
    });
  });
});

describe('<RenderEventBodyInTransactions />', () => {
  const props = {
    cancelingTransactionStarted: false,
    createModalContent: jest.fn(),
    event: sampleEvent,
    startEventCancel: jest.fn(),
  };
  const alterProps = newProps => ({ ...props, ...newProps });
  let wrapper;
  beforeEach(() => {
    wrapper = shallow(<RenderEventBodyInTransactions {...props} />);
  });
  describe('Rendering:', () => {
    it('should render correctly', () => {
      expect(wrapper).toMatchSnapshot();
    });
    it('should render correctly when cancelingTransactionStarted in props is true', () => {
      wrapper.setProps(alterProps({ cancelingTransactionStarted: true }));
      expect(wrapper).toMatchSnapshot();
    });
  });
});

describe('<RenderEventDescription />', () => {
  const props = {
    createModalContent: jest.fn(),
    event: sampleEvent,
  };
  const alterProps = newProps => ({ ...props, ...newProps });
  let wrapper;
  beforeEach(() => {
    wrapper = shallow(<RenderEventDescription {...props} />);
  });
  describe('Rendering:', () => {
    it('should render correctly when description is less than 100 chars', () => {
      expect(wrapper).toMatchSnapshot();
    });
    it('should render truncated text when description is greater 100chars', () => {
      const longDescription = Array.from({ length: 120 }, () => 'x').join('');
      wrapper.setProps(alterProps({ event: { ...props.event, description: longDescription } }));
      expect(wrapper).toMatchSnapshot();
    });
  });
  describe('Interactions:', () => {
    it('should call the function attached to the `Read more` link when description is more that 100chars', () => {
      const longDescription = Array.from({ length: 120 }, () => 'x').join('');
      wrapper.setProps(alterProps({ event: { ...props.event, description: longDescription } }));
      wrapper.find('a[href="#event-details-modal"]').simulate('click');
      expect(props.createModalContent.mock.calls.length).toEqual(1);
    });
  });
});


describe('<RenderEventFooter />', () => {
  const props = {
    edit: jest.fn(),
    event: sampleEvent,
    startDelete: jest.fn(),
  };
  const alterProps = newProps => ({ ...props, ...newProps });
  let wrapper;
  beforeEach(() => {
    wrapper = shallow(<RenderEventFooter {...props} />);
  });
  describe('Rendering:', () => {
    it('should render correctly when event is allowed', () => {
      expect(wrapper).toMatchSnapshot();
    });
    it('should render correctly when event is canceled', () => {
      wrapper.setProps(alterProps({ event: { ...sampleEvent, status: 'canceled' } }));
      expect(wrapper).toMatchSnapshot();
    });
    it('should render correctly when event is done', () => {
      wrapper.setProps(alterProps({ event: { ...sampleEvent, status: 'done' } }));
      expect(wrapper).toMatchSnapshot();
    });
  });
});
