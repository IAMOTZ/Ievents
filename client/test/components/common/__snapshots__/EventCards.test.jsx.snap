// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<EventCards /> Rendering: should render correctly on Events page 1`] = `
<div
  className="col-lg-4 col-md-6 col-sm-1"
  key="1"
>
  <div
    className="card text-dark bg-white mb-3 mx-auto event-card"
    style={
      Object {
        "maxWidth": "20rem",
      }
    }
  >
    <div
      className="card-header"
    >
      <span
        className="h4 mb-0"
      >
        test-event
      </span>
    </div>
    <RenderEventBody
      createModalContent={[MockFunction]}
      event={
        Object {
          "centerId": 1,
          "date": "12/34/2020",
          "description": "test description",
          "id": 1,
          "status": "allowed",
          "title": "test-event",
          "user": Object {
            "email": "tester@gmail.com",
          },
          "userId": 1,
        }
      }
    />
    <RenderEventFooter
      edit={[MockFunction]}
      event={
        Object {
          "centerId": 1,
          "date": "12/34/2020",
          "description": "test description",
          "id": 1,
          "status": "allowed",
          "title": "test-event",
          "user": Object {
            "email": "tester@gmail.com",
          },
          "userId": 1,
        }
      }
      startDelete={[MockFunction]}
    />
  </div>
</div>
`;

exports[`<EventCards /> Rendering: should render correctly on Transactions page 1`] = `
<div
  className="col-lg-4 col-md-6 col-sm-1"
  key="1"
>
  <div
    className="card text-dark bg-white mb-3 mx-auto event-card"
    style={
      Object {
        "maxWidth": "20rem",
      }
    }
  >
    <div
      className="card-header"
    >
      <span
        className="h4 mb-0"
      >
        test-event
      </span>
    </div>
    <RenderEventBodyInTransactions
      cancelingTransactionStarted={false}
      createModalContent={[MockFunction]}
      event={
        Object {
          "centerId": 1,
          "date": "12/34/2020",
          "description": "test description",
          "id": 1,
          "status": "allowed",
          "title": "test-event",
          "user": Object {
            "email": "tester@gmail.com",
          },
          "userId": 1,
        }
      }
      startEventCancel={[MockFunction]}
    />
  </div>
</div>
`;

exports[`<RenderEventBody /> Rendering: should render correctly 1`] = `
<div
  className="card-body"
>
  <div
    className=""
  >
    <RenderEventDescription
      createModalContent={[MockFunction]}
      event={
        Object {
          "centerId": 1,
          "date": "12/34/2020",
          "description": "test description",
          "id": 1,
          "status": "allowed",
          "title": "test-event",
          "user": Object {
            "email": "tester@gmail.com",
          },
          "userId": 1,
        }
      }
    />
    <p
      className="mb-1"
    >
      <i
        className="fa fa-bank fa-fw"
      />
       
      <span
        className="text-capitalize"
      />
    </p>
    <div
      className="date-and-status"
    >
      <span>
        12/34/2020
         
      </span>
    </div>
  </div>
</div>
`;

exports[`<RenderEventBody /> Rendering: should render correctly when the event is not allowed 1`] = `
<div
  className="card-body"
>
  <div
    className="text-muted"
  >
    <RenderEventDescription
      createModalContent={[MockFunction]}
      event={
        Object {
          "centerId": 1,
          "date": "12/34/2020",
          "description": "test description",
          "id": 1,
          "status": "canceled",
          "title": "test-event",
          "user": Object {
            "email": "tester@gmail.com",
          },
          "userId": 1,
        }
      }
    />
    <p
      className="mb-1"
    >
      <i
        className="fa fa-bank fa-fw"
      />
       
      <span
        className="text-capitalize"
      />
    </p>
    <div
      className="date-and-status"
    >
      <span>
        12/34/2020
         
      </span>
    </div>
  </div>
</div>
`;

exports[`<RenderEventBodyInTransactions /> Rendering: should render correctly 1`] = `
<div
  className="card-body"
>
  <RenderEventDescription
    createModalContent={[MockFunction]}
    event={
      Object {
        "centerId": 1,
        "date": "12/34/2020",
        "description": "test description",
        "id": 1,
        "status": "allowed",
        "title": "test-event",
        "user": Object {
          "email": "tester@gmail.com",
        },
        "userId": 1,
      }
    }
  />
  <p
    className="mb-1"
  >
    <i
      className="fa fa-user fa-fw"
    />
     
    <span>
      tester@gmail.com
    </span>
  </p>
  <div
    className="date-and-status"
  >
    <span>
      12/34/2020
       
    </span>
  </div>
  <button
    className="btn ie-red-button mt-2 btn-block"
    data-target="#confirmation-modal"
    data-toggle="modal"
    id={1}
    onClick={[MockFunction]}
  >
    <span>
      Cancel Event
    </span>
  </button>
</div>
`;

exports[`<RenderEventBodyInTransactions /> Rendering: should render correctly when cancelingTransactionStarted in props is true 1`] = `
<div
  className="card-body"
>
  <RenderEventDescription
    createModalContent={[MockFunction]}
    event={
      Object {
        "centerId": 1,
        "date": "12/34/2020",
        "description": "test description",
        "id": 1,
        "status": "allowed",
        "title": "test-event",
        "user": Object {
          "email": "tester@gmail.com",
        },
        "userId": 1,
      }
    }
  />
  <p
    className="mb-1"
  >
    <i
      className="fa fa-user fa-fw"
    />
     
    <span>
      tester@gmail.com
    </span>
  </p>
  <div
    className="date-and-status"
  >
    <span>
      12/34/2020
       
    </span>
  </div>
  <button
    className="btn ie-red-button mt-2 btn-block"
    data-target="#confirmation-modal"
    data-toggle="modal"
    id={1}
    onClick={[MockFunction]}
  >
    <span>
      Canceling 
      <i
        className="fa fa-spinner fa-pulse"
      />
    </span>
  </button>
</div>
`;

exports[`<RenderEventDescription /> Rendering: should render correctly when description is less than 100 chars 1`] = `
<div
  className="description"
>
  <span>
    test description
  </span>
</div>
`;

exports[`<RenderEventDescription /> Rendering: should render truncated text when description is greater 100chars 1`] = `
<div
  className="description"
>
  <span>
    <span>
      xxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxx...
    </span>
    <span
      className="read-more-link"
    >
       
      <a
        className="no-outline-btn"
        data-toggle="modal"
        href="#event-details-modal"
        onClick={[Function]}
        role="button"
      >
        Read more
      </a>
    </span>
  </span>
</div>
`;

exports[`<RenderEventFooter /> Rendering: should render correctly when event is allowed 1`] = `
<div
  className="card-footer"
>
  <div
    className="action-links"
  >
    <span>
      <Link
        id={1}
        onClick={[MockFunction]}
        replace={false}
        to="/events/1/edit"
      >
        <i
          className="fa fa-pencil fa-fw text-dark"
        />
        <span>
          Edit Event
        </span>
      </Link>
    </span>
    <span>
      <a
        data-toggle="modal"
        href="#confirmation-modal"
        id={1}
        onClick={[MockFunction]}
      >
        <i
          className="fa fa-trash text-dark"
        />
         
        <span>
          Delete Event
        </span>
      </a>
    </span>
  </div>
</div>
`;

exports[`<RenderEventFooter /> Rendering: should render correctly when event is canceled 1`] = `
<div
  className="card-footer"
>
  <div
    className="action-links"
  >
    <span>
      <span>
        <i
          className="fa fa-times text-danger"
        />
         
        <span
          className="text-capitalize"
        >
          canceled
        </span>
      </span>
    </span>
    <span>
      <a
        data-toggle="modal"
        href="#confirmation-modal"
        id={1}
        onClick={[MockFunction]}
      >
        <i
          className="fa fa-trash text-dark"
        />
         
        <span>
          Delete Event
        </span>
      </a>
    </span>
  </div>
</div>
`;

exports[`<RenderEventFooter /> Rendering: should render correctly when event is done 1`] = `
<div
  className="card-footer"
>
  <div
    className="action-links"
  >
    <span>
      <span>
        <i
          className="fa fa-check text-success"
        />
         
        <span
          className="text-capitalize"
        >
          done
        </span>
      </span>
    </span>
    <span>
      <a
        data-toggle="modal"
        href="#confirmation-modal"
        id={1}
        onClick={[MockFunction]}
      >
        <i
          className="fa fa-trash text-dark"
        />
         
        <span>
          Delete Event
        </span>
      </a>
    </span>
  </div>
</div>
`;
